python/gink/crud.py

```diff
@@ -0,0 +1,149 @@
+"""
+The WSGI API application for gink operations.
+
+To run this application, pass --wsgi gink.crud.app when starting gink.
+
+Set auth key with env AUTH_TOKEN.
+
+"""
+
+from json import loads, dumps
+from os import environ
+
+from .impl.directory import Directory
+
+
+class Crud():
+    """
+    WSGI application to GET, PUT, and DELETE data in a Gink database.
+
+    Example for PUT:
+
+    curl -X PUT http://localhost:8099/key -d 3 -H "Content-Type: application/json"
+
+    # Sets "key" in root directory to 3
+
+    Example for GET:
+
+    curl -X GET http://localhost:8099/key
+
+    # Returns 3
+
+    """
+    def __init__(self):
+        self.root = Directory(arche=True)
+        self.auth_token = environ.get("AUTH_TOKEN")
+
+    def __call__(self, environ, start_response):
+        # If auth token is present, expect token in the Authorization header.
+        if self.auth_token:
+            auth_header = environ.get('HTTP_AUTHORIZATION')
+            if not auth_header or self.auth_token not in auth_header:
+                return self._bad_auth_handler(start_response)
+
+        raw_path = environ.get('PATH_INFO')
+        if not raw_path or raw_path == '/':
+            return self._bad_path_handler(start_response)
+
+        if environ.get("REQUEST_METHOD") == "GET":
+            default = object()
+            result = self.root.get(raw_path.split("/"), default)
+            if result is default:
+                return self._data_not_found_handler(start_response)
+            return self._get_handler(result, start_response)
+
+        elif environ.get("REQUEST_METHOD") == "PUT":
+            request_body: bytes = environ.get("wsgi.input").read()
+            if not request_body:
+                return self._bad_body_handler(start_response)
+
+            content_type = environ.get("HTTP_CONTENT_TYPE")
+            if content_type == "application/octet-stream" or content_type == "application/x-www-form-urlencoded":
```

Comment:
Though I think that defaulting to assuming binary data if no content-type header is sent (i.e. you content_type variable being None or empty), I don't think that's the same as treating form encoded data as binary as binary data (though on the other hand maybe it's just more practical/expedient to follow that behavior because it's the default that curl uses).

Reply:
This is what I landed on:

```py
content_type = env.get("HTTP_CONTENT_TYPE")
        if not request_body:
            value = None
        elif content_type == "text/plain":
            value = request_body.decode()
        elif content_type == "application/json":
            value = loads(request_body.decode())
        else: # includes octet-stream and x-www-form-urlencoded
            value = request_body # Default to binary
```

Does that work?